Code for 'Innovations in Spinal Cord Cell Type Heterogeneity Across Vertebrate Evolution' by *Ignatyev et al.*, 2025 (https://www.biorxiv.org/content/10.1101/2025.10.09.680955v2)

Datasets are currently available via private link and will be made available upon publication. The repository will be updated with additional scripts as needed. 

---

## üìò Overview
This repository contains the analysis scripts and notebooks used to produce the results and figures in the manuscript.  
The study integrates developmental and adult single-cell/nucleus and spatial transcriptomic data from **Xenopus laevis**, **mouse**, **human**, and **zebrafish** to explore the evolutionary origins of spinal cord neuronal diversity.

![](abstract.jpg)

---

## üìÅ Structure
The repository is organized into subdirectories corresponding to the analyses used to generate the figures:

- **scripts/**
  - Contains codes used for data processing and cross-species analyses.
  - Analyses are grouped into:
    - **Development** (Figures **1‚Äì2**, Supplementary **S1‚ÄìS9**) <br>
      ‚Ä¢ data processing (01.-03.) <br>
      ‚Ä¢ transforming frog data into mouse format (04. - was also used for transforming **adult** frog data) <br>
      ‚Ä¢ purity analysis and Pearson correlation of cardinal class expression profiles (05.) - Fig.2D, Supp4 <br>
      ‚Ä¢ developmental integration (06.) - Fig.2E, Supp3 <br>
      ‚Ä¢ similarity index in development (07.), Fig.2E, Supp3 <br>
      ‚Ä¢ frog spatial data analysis (08.), Fig.2F-I, Supp7-9 <br>
      
    - **Adult** (Figures **3‚Äì4**, Supplementary **S11‚ÄìS18**) <br>
      ‚Ä¢ frog adult snrna-seq data processing  (01.) <br> -  Fig.3
      ‚Ä¢ frog adult spatial data processing (02.) <br> - Fig.3
      ‚Ä¢ adult integration (03.) <br>
      ‚Ä¢ similarity index in adult (04.) - Fig.4F <br>
      ‚Ä¢ species mixing (05.), Fig.4, Supp18 <br>
      ‚Ä¢ projecting divergent and conserved populations onto frog and mouse spatial data (06.), Fig.4J
    
---

## Dependencies

The following core R packages were used:

- **Seurat**
- **dplyr**
- **ggplot2**
- **tidyr**
- **data.table**
- **Matrix**
- **patchwork**
- **cowplot**
- **mixtools**
- **tibble**
- **biomaRt**
- **stringr**
- **grid**
- **purrr**
- **sp**
- **ggtern**
- **ggrepel**
- **viridis**
- **RColorBrewer**
- **gridExtra**
- **future**
- **future.apply**
- **ComplexHeatmap**
- **CellChat**
- **circlize**



